File: day:25/sortedArrayToBST.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def sortedArrayToBST(self, nums: List[int]) -> TreeNode:
        #base cases
        if(len(nums) == 0): return None;
        if(len(nums) == 1): return TreeNode(nums[0]);
        
        #find the mid of the array
        mid = len(nums)//2
        left = nums[0:mid]
        right = nums[mid + 1:]
    
        node = TreeNode(nums[mid]);
        node.left = self.sortedArrayToBST(left);
        node.right = self.sortedArrayToBST(right);
        return node;
        

File: day:25/sortedArrayToBST.kt
/**
 * Example:
 * var ti = TreeNode(5)
 * var v = ti.`val`
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */
class Solution {
    fun sortedArrayToBST(nums: IntArray): TreeNode? {
        // base cases
        if(nums.size == 0) return null;
        if(nums.size == 1) {
            return TreeNode(nums[0]);
        }
        // find the mid of the array
        var mid: Int = Math.floor(nums.size/2.toDouble()).toInt();
        var left = nums.sliceArray(0..mid-1);
        var right = nums.sliceArray(mid + 1..nums.size - 1);
    
        var node = TreeNode(nums[mid]);
        node.left = sortedArrayToBST(left);
        node.right = sortedArrayToBST(right);
        return node;
    }
}
